Cart

adam-adair opened the pull request:
This proved harder than I thought, so there's maybe too much going on for one PR. Please review carefully!
But anyway, you can add items to your cart as a guest or signed in user, and it will handle combining and adding to the DB appropriately.
This also has auth middleware for making sure you're making changes to the right cart.

adam-adair committed:
Merge branch 'products-view-updates' of github.com:FullstackAcademy/2007-class-acts into minor-fixes

adam-adair committed:
merge dev into cart

adam-adair committed:
Merge branch 'dev' of github.com:FullstackAcademy/2007-class-acts into cart

adam-adair committed:
some frontend cart features

adam-adair committed:
partial cart backend

adam-adair committed:
partial cart

adam-adair committed:
logged in user cartItem logic

adam-adair committed:
partial local cart implementation

adam-adair committed:
can view items in cart component, handles login/user creation

adam-adair committed:
fixed multiple carts bug

zoegreene commented on client/components/Account.js:
```diff
@@ -14,8 +14,12 @@ const Account = ({ user, _destroySession }) => {
-          _destroySession(sessionId)
+          //get rid of the session so you're not automatically logged back in
+          _destroySession(sessionId);
+          //empty your cart if you log out
```
Perhaps this is semantics, but this comment is a little unclear. The setCartItems will reset the current state of the cart to [], but when you log back in as this user you still have items in the cart. Not sure if this was the intended behavior, but I think this is the correct behavior regardless.

adam-adair replied:
Yeah, I meant for this to empty the front-end, redux cart, so that you don't have everything that was in your basket still there after you've logged out (but it is in the DB, associated w your account).
Mostly I was trying to avoid having stuff in your cart when you log out that then gets added _again_ to your cart when you log back in. I figure if the cart is empty after logging out, it's clear enough that that stuff was your "account cart" only.

zoegreene commented on client/components/Cart.js:
```diff
@@ -0,0 +1,47 @@
+    //if you go directly to the cart page, you don't have the artwork, so this is a workaround but we probably need to figure out the right way to do this later
+    if(artworks.length === 0) {
+      this.props.getArtworks();
+      return (<div>Loading</div>)
```
If you have 0 artworks, even after you getArtworks it shows a blank screen. Should add a catch here that says "No items in cart" or something.

adam-adair replied:
This isn't the cart artwork, this is *all* the artwork. I'm just getting it here so that I can look up stuff like title, artist, pic, etc., in order to render a nice cart. The cart only stores the artworkId and quantity (which was probably a mistake).

zoegreene commented on client/components/Cart.js:
```diff
@@ -0,0 +1,47 @@
+    //this is crummy looking but works for viewing the cart for now
+    return (
+      <div>
+        { cart.map(cartItem => {
```
Same comment here as above - if nothing in the cart page is blank, but should have some boilerplate text.

adam-adair replied:
Agreed

adam-adair replied:
done

zoegreene commented on client/components/Cart.js:
```diff
@@ -0,0 +1,47 @@
+              <h4>{ cartItemArtwork.title }</h4>
+              <h5><i>{ cartItemArtwork.artist.name }</i></h5>
+              <h5>
+                Quantity: { cartItem.quantity }
```
We should have an X button next to each item or a drop-down next to the quantity so that users can remove items from the cart. Currently there's no way to do that without clearing your cookies.

adam-adair replied:
Yeah definitely. You actually can't remove them at all if you've got a logged in cart.

adam-adair replied:
done, in theory

zoegreene commented on client/components/LoginScreen.js:
```diff
@@ -24,10 +25,15 @@ class LoginScreen extends Component {
       const res = await axios.post('/api/users/login', { ...this.state })
       const user = res.data
```
Could make this a one-liner with 24

adam-adair replied:
aren't we fancy

zoegreene commented on client/components/NewUser.js:
```diff
@@ -23,10 +24,13 @@ class NewUser extends Component {
       const res = await axios.post('/api/users', { ...this.state })
       const user = res.data
```
Same comment as before... can make this into a one-liner

zoegreene commented on client/components/SingleArtwork.js:
```diff
@@ -1,49 +1,69 @@
+    }
+
+    async addQty(){
+        const quantity = +document.getElementById('qty').value
```
I think you should be able to get the ev.target.value instead of having to use getElementById here

adam-adair replied:
The event here would be the button press. I could store qty in the state, but this seemed simpler.

zoegreene commented on client/components/SingleArtwork.js:
```diff
@@ -1,49 +1,69 @@
-                :
-
-                <h2>No image found</h2>}
+                <div style={{padding: "1rem"}}>
```
Lest we forget, this should be moved to the style.css file

zoegreene commented on server/api/cart.js:
```diff
@@ -0,0 +1,54 @@
+const express = require('express')
```
For hygiene, please add next and try/catch blocks to these

zoegreene commented on server/api/cart.js:
```diff
@@ -0,0 +1,54 @@
+const router = express.Router()
+const { Cart, CartItem } = require('../db')
+
+router.post('/item', async (req, res) => {
```
Feels like these two POST routes could be DRYed with some logic

adam-adair replied:
Probably so. Might be able to make it call the DB fewer times, but might obscure the 'flow' of this route, though, too. I don't have any great ideas.

zoegreene commented on server/api/users.js:
```diff
@@ -1,13 +1,24 @@
+const { User, Session, Cart, CartItem } = require('../db')
 const bcrypt = require('bcrypt')
 
 const A_WEEK_IN_SECONDS = 60 * 60 * 24 * 7;
```
Add to constants file?

adam-adair replied:
We could. This is the only place it's used, though.

zoegreene commented on server/api/users.js:
```diff
@@ -1,13 +1,24 @@
-    const session = await Session.findByPk(req.params.sessionId, { include: [User] })
+    const session = await Session.findByPk(req.params.sessionId, {
+      //we need to include the cart info w the user on login and on session recognition
+      include: [
```
Holy mother of includes

adam-adair replied:
kill me

zoegreene reviewed (commented):
Holy moly this was a lot! I think the logic here holds and I've tested it out a bit from localhost. Mostly comments here about how to improve the UX/functionality and code hygiene. Thanks for building out this structure! We've still got a bit to work on in terms of error handling, admin functionality, etc.

adam-adair commented:
> Holy moly this was a lot! I think the logic here holds and I've tested it out a bit from localhost. Mostly comments here about how to improve the UX/functionality and code hygiene. Thanks for building out this structure! We've still got a bit to work on in terms of error handling, admin functionality, etc.

I couldn't figure out how to reply directly under your comment, which is sort of annoying.

Anyway, yeah, it really is a lot. Started to feel a bit spaghetti-fied, but I wasn't sure how to get around that. Cart logic is weird.

I'll work on addressing your comments and push an update. Would love other class actors' feedback.

jamilkabir commented:
Hey All. I’ve had a migraine since last night and finally starting to feel better. I’ll send feedback on it tonight, and also have user section of the admin portal completed. 

Also, the internet in my area is out. I’ll be available on slack in case the internet isn’t back on. 

> On Oct 27, 2020, at 3:53 PM, Adam Reeves <notifications@github.com> wrote:
> 
> ﻿
> Holy moly this was a lot! I think the logic here holds and I've tested it out a bit from localhost. Mostly comments here about how to improve the UX/functionality and code hygiene. Thanks for building out this structure! We've still got a bit to work on in terms of error handling, admin functionality, etc.
> 
> I couldn't figure out how to reply directly under your comment, which is sort of annoying.
> 
> Anyway, yeah, it really is a lot. Started to feel a bit spaghetti-fied, but I wasn't sure how to get around that. Cart logic is weird.
> 
> I'll work on addressing your comments and push an update. Would love other class actors' feedback.
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or unsubscribe.

adam-adair committed:
minor fixes and comments

adam-adair committed:
local implementation of cart item qty change/delete

adam-adair committed:
implement cartitem change/delete backend

adam-adair commented:
I think I've addressed the higher priority comments and added the quantity changes/item deletion with most recent commit and push. There are some things that could be nicer, but I *think* this is a fully functional cart now. Again, would be great if everybody could review it, because it feels like a lot going on.

z-rodney commented on client/components/Cart.js:
```diff
@@ -0,0 +1,98 @@
+    removeCartItem: (cartItem, isLoggedIn) => dispatch(removeCartItem(cartItem, isLoggedIn))
+  }
+}
+export default connect(mapStateToProps, mapDispatchToProps)(Cart)
```
this logic all seems to make sense. this is relatively minor but this.props.getArtworks might be better in a component did mount, and the defined functions might be better as methods on the component? i'm not 100% sure what's the best practice.

adam-adair replied:
Yeah I agree, should all be component methods. I don't know what the difference is functionally, either, but I'll do it the way we learnt it.

adam-adair replied:
done in latest commit

z-rodney commented on client/components/SingleArtwork.js:
```diff
@@ -1,49 +1,69 @@
+        this.addQty = this.addQty.bind(this)
+    }
+
+    async addQty(){
```
don't think this async needs to be here

adam-adair replied:
doh

z-rodney commented on client/components/SingleArtwork.js:
```diff
@@ -1,49 +1,69 @@
+        const quantitySelection = []
+        for (let i = 1; i <= this.props.artwork.quantity; i++) {
+            quantitySelection.push(<option key={i} value={i}>{i}</option>)
+        }
```
i understand the logic here, one thing i'm considering is that for artwork that has A LOT in stock (for ex, one piece in the seed file has a quantity over 700) i'm not sure if it makes sense for this to be a dropdown. this is more so business logic, but if hypothetically people aren't mass buying art then we can probably just limit it to 10 or quantity, whichever is less.

adam-adair replied:
Yeah I thought about that, too. I think the way you describe it (drop down to 10, manually enter >10) is the way Amazon does it. Maybe for the purposes of this class, we should limit our inventory to 10...

z-rodney commented on client/redux/cart.js:
```diff
@@ -0,0 +1,119 @@
+}
+
+// THUNK CREATORS
+export const addCartItem = (cartItem, isLoggedIn = false) => {
```
since by default isLoggedIn is false, i think that means you don't have to pass in isLoggedIn if it isn't true in the cart component, lines 34 - 44.

adam-adair replied:
That's true, but when these thunk creators are called in the Cart component or SingleArt component, I don't know whether it's true or not.

z-rodney commented on client/components/Cart.js:
```diff
@@ -0,0 +1,98 @@
+      if(this.props.user.id) isLoggedIn = true
+      this.props.removeCartItem(cartItem, isLoggedIn)
+    }
+
```
made this comment in the redux/cart file, but you dont need to pass isLoggedIn every time if it isn't true, since by default it's false.

adam-adair replied:
I responded in redux/cart, but I'll say a little more here: lines 41 and 42 set isLoggedIn to true or false, which needs to be passed to removeCartItem on 43. I don't think it could be left out.

z-rodney commented on client/redux/cart.js:
```diff
@@ -0,0 +1,119 @@
+    //then dispatch the original cart item for the reducer
+    if(deleteSuccessful === 204) {
+      dispatch(_removeCartItem(cartItem));
+    }
```
i think try catch would be useful here

adam-adair replied:
Yeah. Probably all of these should have some kind of "oops, sorry, we can't do that to your cart right now" error handling. Or did you have something else in mind for try/catch?

z-rodney commented on client/redux/user.js:
```diff
@@ -13,6 +14,7 @@ export const getUser = (sessionId) => {
   return async (dispatch) => {
     const user = await axios.get(`/api/users/${sessionId}`)
     dispatch(setUser(user.data))
+    if(user.data.cart) dispatch(setCartItems(user.data.cart.cartItems))
```
i understand the logic, but i think this needs to be moved to the cart store instead. so in the cart store you would account for action type getUser and if there's cart data then handle it there.

adam-adair replied:
Ahhh really good idea. This felt gross, but I couldn't think of the right way.

adam-adair replied:
done in commit

z-rodney commented on server/api/cart.js:
```diff
@@ -0,0 +1,153 @@
+          cartId: cart.id
+        },
+        returning: true, // needed for cartItems to be populated
+        plain: true // makes sure that the returned instances are just plain objects
```
i've found these updates to be trickier than expected. another approach i've used to get around this is finding the item first then updating, more DB calls but simpler code.

adam-adair replied:
Part of the reason I did it this way is that it needs to make sure the cart belongs to the user making the request. For this route, at least, I think you could go directly to the cartItem using the ID and probably include the cart and include the user, but that seems equally tricky to me. Plus, it wouldn't work for adding cartItems, since there are ones that come from localStorage that do not yet have an ID (although it could be argued that the current way cartItems are added is kinda hacky), so I saw some value in consistency between routes.
Probably it was a mistake to make carts at all, since a user really only should have one cart. If we change that and change the hacky way cartItems are added, it could simplify all these routes. Feels like a lot of backtracking, though. Maybe we can discuss in class tonight.

z-rodney reviewed (commented):
overall it looks pretty good! i think it would be helpful to fully talk through all of this in class tomorrow so we can talk through anything we might need to change to set us up for orders.

jamilkabir commented on client/components/AllArtwork.js:
```diff
@@ -33,18 +37,16 @@ export class AllArtwork extends Component {
     });
   }
+
   async changeFilter(ev) {
```
Can we add a comment on how this function works. Also renaming the function to something like filtersByCategoryFromDropDown would make it more clear, though its a long variable name haha

adam-adair replied:
this ain't really part of this PR

jamilkabir reviewed (commented):
Looks good.

adam-adair committed:
fixes per Zaina PR comments

z-rodney reviewed (approved):
looking good!

z-rodney committed:
Merge branch 'dev' into cart
